File: hs/src/Day25.hs
{-|
Module:         Day25
Description:    <https://adventofcode.com/2021/day/25 Day 25: Sea Cucumber>
-}
{-# LANGUAGE FlexibleContexts, TypeApplications #-}
module Day25 (day25) where

import Control.Arrow (first, second)
import Control.Monad (guard)
import Data.Array.Unboxed (UArray, (!), (//), accumArray, assocs)
import Data.Maybe (isJust)
import Data.Text (Text)
import qualified Data.Text as T (length, lines, unpack)

day25 :: Text -> Maybe Int
day25 input = do
    let inputs = T.lines input
        height = length inputs
    width <- maximum $ Nothing : map (Just . T.length) inputs
    let state0 = accumArray @UArray (const id) '.' ((0, 0), (height - 1, width - 1))
            [((y, x), c) | (y, line) <- zip [0..] inputs, (x, c) <- zip [0..] $ T.unpack line]
        step state = state'' <$ guard (a || b) where
            (a, state') = step' (second $ \x -> succ x `mod` width) '>' state
            (b, state'') = step' (first $ \y -> succ y `mod` height) 'v' state'
        step' f d state = (not $ null acc, state // acc) where
            acc = do
                (i, c) <- assocs state
                guard $ c == d && state ! f i == '.'
                [(i, '.'), (f i, d)]
    pure $ length $ takeWhile isJust $ iterate (>>= step) $ Just state0

File: kt/src/commonMain/kotlin/com/github/ephemient/aoc2021/Day25.kt
package com.github.ephemient.aoc2021

/** [Day 25](https://adventofcode.com/2022/day/25): Sea cucumber */
class Day25(lines: List<String>) {
    private val height = lines.size
    private val width = lines.maxOfOrNull { it.length } ?: 0
    private val initialState = CharArray(width * height) { lines[it / width].getOrElse(it % width) { '.' } }

    fun part1(): Int {
        var n = 1
        val state = initialState.copyOf()
        while (true) {
            val a = (0 until height).fold(true) { acc, y ->
                val indices = (0 until width).filter { x ->
                    state[y * width + x] == '>' && state[y * width + (x + 1) % width] == '.'
                }
                for (x in indices) {
                    state[y * width + x] = '.'
                    state[y * width + (x + 1) % width] = '>'
                }
                acc && indices.isEmpty()
            }
            val b = (0 until width).fold(true) { acc, x ->
                val indices = (0 until height).filter { y ->
                    state[y * width + x] == 'v' && state[(y + 1) % height * width + x] == '.'
                }
                for (y in indices) {
                    state[y * width + x] = '.'
                    state[(y + 1) % height * width + x] = 'v'
                }
                acc && indices.isEmpty()
            }
            if (a && b) return n else n++
        }
    }
}
